examples/logisticregression/main.go

```diff
@@ -1,185 +1,288 @@
+	// "dcost/dw" == derivative of cost with respect to w
+	G.WithName("dcost/dw")(grads[0])
+	// "dcost/db" == derivative of cost with respect to b
+	G.WithName("dcost/db")(grads[1])
+
+	// create the nodes for calculating the gradient
+	learnRate := G.NewConstant(0.1) // be careful not to set a learnRate too high
+	gwlr := G.Must(G.Mul(learnRate, grads[0]))
+	wUpd := G.Must(G.Sub(w, gwlr))
+	gblr := G.Must(G.Mul(learnRate, grads[1]))
+	bUpd := G.Must(G.Sub(b, gblr))
+
+	// create the nodes that do the final operation of updating w and b
+	G.Set(w, wUpd)
+	G.Set(b, bUpd)
+	// write to the gographviz file for debugging https://github.com/awalterschulze/gographviz
+	ioutil.WriteFile("fullGraph.dot", []byte(g.ToDot()), 0644)
+
+	// Now that we have created all the notes, we can compile a program.
+	// We are essentially creating a list of instructions to get from our inputs {x, y}
+	// to our outputs {wUpd, bUpd}. Note that we only need to create the progam once.
+	prog, locMap, err := G.CompileFunction(g, G.Nodes{x, y}, G.Nodes{wUpd, bUpd})
+	handleError(err)
+	fmt.Printf("%v", prog) // print the instructions
+	// With our program, we initialize a new TapeMachine that will execute our program
+	machine := G.NewTapeMachine(g, G.WithPrecompiled(prog, locMap))
+
+	// we allocated the node w before, but we never set it's initial value
+	machine.Let(w, wT)
+	machine.Let(b, 0.0)
+	// ?? Why do we need to profile the CPU?
+	if *cpuprofile != "" {
+		f, err := os.Create(*cpuprofile)
+		if err != nil {
+			log.Fatal(err)
+		}
+		pprof.StartCPUProfile(f)
+		defer pprof.StopCPUProfile()
+	}
+
+	// Now that we have our graph, program, and machine, we can start training
+	start := time.Now()
+	for i := 0; i < trainIter; i++ {
+		// move the pointer back to the beginning of the prog. Reset() does not delete any values
+		machine.Reset()
+		// ?? why do we need to reinitialize the values the nodes {x,y}?
+		machine.Let(x, xT)
+		machine.Let(y, yT)
+		handleError(machine.RunAll())
+		// ?? Shouldn't the program have already done this? The last node has this Set() operation
+		//	G.Set(w, wUpd)
+		//	G.Set(b, bUpd)
+		machine.Set(w, wUpd)
+		machine.Set(b, bUpd)
+	}
+	fmt.Printf("Time taken: %v\n", time.Since(start))
+	fmt.Printf("Final Model: \nw: %3.3s\nb: %+3.3s\n", w.Value(), b.Value())
+
+	fmt.Printf("Target values: %#v\n", yT)
+	fmt.Printf("START\n")
+
+	// Now that we have our final model, we need to write a new program
+	// that goes from the input {x} to the prediction {pred}
+	prog, locMap, err = G.CompileFunction(g, G.Nodes{x}, G.Nodes{pred})
+	fmt.Printf("%+v", err)
+	handleError(err)
+	machine = G.NewTapeMachine(g, G.WithPrecompiled(prog, locMap))
+
+	// ?? Don't we want to use the weight and bias that we trained? not reset them?
+	machine.Let(w, wT)
+	machine.Let(b, 0.0)
+	// ?? Shouldn't we create a new tensor xT so that we can predict?
+	machine.Let(x, xT)
+	handleError(machine.RunAll())
+	fmt.Printf("Predicted: %#v\n", pred.Value())

```

Comment:
There's one thing that used to be there but somehow amongst the merges, it seems to have been lost - it'd be great if you could add it back in. Here's the type signature:

```
func accuracy(target, predicted G.Value) float64 
```

It should be trivial, - since the predicted and the target has the same shape, just iterate through and divide by. The correct answer for this program should be about 97% accuracy

Reply:
Sure, I can work on this. I was originally going to do an accuracy calculation in _test.go. But I think it does make more sense in here.

Reply:
+1